### PR TITLE
Master dynamic value nby

### PR DESCRIPTION
**Pull request by Goaman**

### CODE REVIEW

File: packages/core/src/VNodes/VNode.ts [hunk 1]
```diff
@@ -104,6 +114,21 @@ export class VNode {
...
+        return this.children
+            .filter(node => node.type !== VNodeType.MARKER)
+            .map(child => {
+                const clonedNode = child.clone();
+                const clonedChildren = child.cloneChildrenDeep();
+                clonedNode.empty();
+                clonedNode.append(...clonedChildren);
+                return clonedNode;
+            });
+    }
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/src/VNodes/VNode.ts:130]: move to fieldnode

File: packages/core/src/VNodes/VNode.ts [hunk 2]
```diff
@@ -648,6 +673,9 @@ export class VNode {
         for (const child of children) {
             this._insertAtIndex(child, 0);
         }
+        if (!children.every(child => child.type === 'markerNode')) {
+            this.fire('childList', { children });
+        }
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/src/VNodes/VNode.ts:678]: remove this

File: packages/core/src/VNodes/VNode.ts [hunk 3]
```diff
@@ -683,13 +717,16 @@ export class VNode {
...
     }
     /**
      * Remove all children of this VNode.
      */
     empty(): void {
-        for (const child of this.children) {
-            child.remove();
+        while (this.children.length) {
+            this.children[0].remove();
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/core/src/VNodes/VNode.ts:729]: add test

File: packages/plugin-odoo-field/src/OdooFieldDomParser.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+     *
+     */
+    _reactiveToFields = new Map<ReactiveValue<any>, Set<OdooFieldNode>>();
+    /**
+     * When a field *node* change, we save the event in this map. This allows to
+     * update all nodes associated to the `ReactiveValue` originating from that
+     * particular *node*. See method _beforeRenderInEditable
+     *
+     */
+    reactiveChanges = new Map<ReactiveValue<any>, OdooFieldNode>();
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/plugin-odoo-field/src/OdooFieldDomParser.ts:78]: try again

File: packages/plugin-odoo-field/src/OdooFieldDomParser.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        const record: OdooRecordDefinition = {
+            modelId: element.attributes['data-oe-model'].value,
+            recordId: element.attributes['data-oe-id'].value,
+            fieldName: element.attributes['data-oe-field'].value,
+        };
+
+        // ? dmo: Same here. What am I supposed to use to get the plugin from a
+        // parser? Or should I have the logic located in the OdooField plugin
+        // instead?
+        const reactivePlugin = this.engine.editor.plugins.get(OdooReactiveRegistry);
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/plugin-odoo-field/src/OdooFieldDomParser.ts:99]: Odoo-only

File: packages/plugin-odoo-field/src/OdooFieldDomParser.ts [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        const reactiveValue = reactivePlugin.getReactiveField(record);
+        const isValidReactive = new ReactiveValue(false);
+
+        const fieldType = element.attributes['data-oe-type'].value;
+        let value = element.innerHTML;
+        let childNodeToParse = element.childNodes;
+        let fieldNode: OdooFieldNode;
+
+        // ? dmo: Special case 1/4 for monetary field. Should I place the logic
+        // specific for monetary somewhere else?
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/plugin-odoo-field/src/OdooFieldDomParser.ts:109]: try again

File: packages/plugin-odoo-field/src/OdooFieldDomParser.ts [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        // Update the ReactiveValue "valid" automatically when the ReactiveValue
+        // reprensenting the field changes.
+        const isValidCallback = (): void => {
+            isValidReactive.set(isFieldValid(fieldType, reactiveValue.get()));
+        };
+        reactiveValue.on('set', isValidCallback);
+
+        const fieldSet = this._reactiveToFields.get(reactiveValue);
+        if (!fieldSet) {
+            reactiveValue.set(value);
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/plugin-odoo-field/src/OdooFieldDomParser.ts:149]: try again, with informing the "odoo reactive registry"

File: packages/plugin-odoo-field/src/OdooFieldDomParser.ts [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+    /**
+     * Before the Dom plugin render, make a deep clone of the field nodes that
+     * needs to be.
+     *
+     * @private
+     * @returns {Promise<void>}
+     * @memberof OdooFieldDomParser
+     */
+    private async _beforeRenderInEditable(): Promise<void> {
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/plugin-odoo-field/src/OdooFieldDomParser.ts:175]: The semantic is wrong to render "before render". Try again.

File: packages/plugin-odoo-field/src/OdooFieldDomParser.ts [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+
+            // Special case 3/4 for monetary field.
+            if (currencyElement) {
+                value = currencyElement.textContent;
+            }
+            await reactiveValue.set(value);
+
+            for (const field of this._reactiveToFields.get(reactiveValue)) {
+                if (fieldNodeToCopy === field) continue;
+                const childs = fieldNodeToCopy.cloneChildrenDeep();
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/plugin-odoo-field/src/OdooFieldDomParser.ts:197]: `cloneChildrenDeep` only on this class

File: packages/plugin-odoo-reactive-registry/src/OdooReactiveRegistry.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
+import { JWPlugin, JWPluginConfig } from '../../core/src/JWPlugin';
+import JWEditor from '../../core/src/JWEditor';
+import { ReactiveValue } from '../../utils/src/ReactiveValue';
+
+/**
+ * The record definion is mainly used to get a reactive field using the method `getReactiveField`.
+ *
+ * @export
+ * @interface OdooRecordDefinition
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/plugin-odoo-reactive-registry/src/OdooReactiveRegistry.ts:9]: remove them

[the remark this answers is not in the document]
Comment (Goaman) [reply packages/plugin-odoo-reactive-registry/src/OdooReactiveRegistry.ts:9]: and deactivate the plugin

File: packages/plugin-odoo-reactive-registry/src/OdooReactiveRegistry.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    modelId: string;
+    recordId: string;
+    fieldName: string;
+}
+
+export class OdooReactiveRegistry<T extends JWPluginConfig = JWPluginConfig> extends JWPlugin<T> {
+    registry: Record<string, ReactiveValue<any>> = {};
+    constructor(public editor: JWEditor, public configuration: T) {
+        super(editor, configuration);
+    }
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/plugin-odoo-reactive-registry/src/OdooReactiveRegistry.ts:21]: remo

File: packages/utils/src/VEvent.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
+/**
+ * Abstract class to add event mechanism.
+ *
+ * @export
+ * @class VEvent
+ * @template T
+ */
+export class VEvent<T extends VEvent<T>> {
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/utils/src/VEvent.ts:8]: EventMixin

File: packages/utils/src/VEvent.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    async fire<T>(eventName: string, args?: T): Promise<void> {
+        if (this._eventCallbacks[eventName]) {
+            for (const callback of this._eventCallbacks[eventName]) {
+                await callback(args);
+            }
+        }
+        if (this.parent) {
+            await this.parent.fire(eventName, args);
+        }
```
[the remark this answers is not in the document]
Comment (Goaman) [reply packages/utils/src/VEvent.ts:41]: nope

[the remark this answers is not in the document]
Comment (Goaman) [reply packages/utils/src/VEvent.ts]: Actually yes, in the field HTML

**Comment by dmo-odoo** [comment]
I'm marking this PR as `needs rebase` because https://github.com/odoo-dev/jabberwock/pull/163 is more critical and will be merged first, requiring the renderers of this PR to be adapted accordingly.

**Comment by dmo-odoo** [comment]
In order to remove the `nextRenders` and `beforeInEditable` hooking stuff, I had to change the content of the `value` of the html field. Writing in an html field mirrors the changes as expected, but saving it in Odoo will require to call `editor.render` instead of just reading its `fieldInfo.value.get()` as the `value` does not contain the corresponding html rendering anymore. I know this is somewhat wrong, but it can be worked around easily IMO by calling `render` when you need the *actual* value and I think it introduces less complexity in the editor code than the previous version with the async hooks. We will probably need to revisit this in the future, though.

Note: setting modifiers do not trigger any event so that is not mirrored yet, though it is unrelated to the change I mentioned above.

**Pull request merged**